### ISSUE TITLE
NLL:  cast causes failure to promote to static

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#55288]
spawned off of https://github.com/rust-lang/rust/issues/55223#issuecomment-431889938

This example is causing an error in NLL:

[play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=3a4a5bcd339c3974fd41aa1e42b1bfc2)

```rust
struct Slice(&'static [&'static [u8]]);
static MAP: Slice = Slice(&[
    b"CloseEvent" as &'static [u8],
]);
```

but this one does not cause an error:

[play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=acb604c330bb9aace7cbfc9d29b529d0)

```rust
struct Slice(&'static [&'static [u8]]);
static MAP: Slice = Slice(&[
    b"CloseEvent",
]);
```

**Comment by estebank** [comment]
And it causes an ICE without nll enabled.

**Comment by oli-obk** [comment]
@davidtwco and I have worked out a theory on zulip:

Promotion is happening successfully, but the `AscribeUserTy` still refers to the old value after promotion is done, because https://github.com/rust-lang/rust/blob/master/src/librustc_mir/transform/promote_consts.rs#L306 only modifies the assignment, leaving the `AscribeUserTy` in place. We need to find potential corresponding `AscribeUserTy`s and update them.
